ensure proper encoding of url and signed values

pgerber opened the pull request:
* Use percent encoding as [specified by Amazon](https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html).
* Use canonical path in URL for requests. Using '+' never worked on
  Amazon and [stopped working for Ceph](https://github.com/ceph/ceph/commit/20e5ff023ebad89c386a520d07613547d4836399). Encoding the whole URL
  should avoid similar issues in the future
* Don't `String::to_ascii_lowercase` when getting canonical headers.
  Keys are already lower cased when inserted.

pgerber committed:
ensure proper encoding of url and signed values

• Use percent encoding as specified by Amazon [1].
• Use canonical path in URL for requests. Using '+' never worked on
  Amazon and stopped working after [2] for Ceph. Encoding the whole URL
  should avoid similar issues in the future
• Don't `String::to_ascii_lowercase` when getting canonical headers.
  Keys are already lower cased when inserted.

[1]: https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html
[2]: https://github.com/ceph/ceph/commit/20e5ff023ebad89c386a520d07613547d4836399

hanscj1 commented:
Will test now. thx

hanscj1 commented:
Compiled and ran against example 's3.rs' and it failed in a number of areas due to signature calculations:

cargo run --example s3

This assumes your aws credentials are in ~/.aws/credentials

Can you verify?

Thanks

pgerber commented:
Cool, overlooked the examples. I'll try to sort out the signature mismatches.

hanscj1 commented:
Look forward to testing them!!

On Tue, Apr 11, 2017 at 4:18 PM, Peter Gerber <notifications@github.com>
wrote:

> Cool, overlooked the examples. I'll try to sort out the signature
> mismatches.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/lambdastackio/aws-sdk-rust/pull/10#issuecomment-293387908>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAhgSAqQYl_5Wbr3w8UVmpYVs2hgkOVJks5ru-ArgaJpZM4M6YVD>
> .
>

pgerber committed:
deal correctly with ?… parameters in URLs

• Only encode the object key and not the whole path. V2 implementation
  adds parameters like ?versioning and we don't want the ? encoded.
• Switch back from canonical_uri to path. Object keys are now correctly
  encoded in the path.

pgerber committed:
use GNU utility compatible arguments for `dd`

• BSD's version of dd appears to expect a lower case 'm' for Megabyte
  and GNU's a capital 'M'.
• verify dd succeeded

pgerber commented:
Still see this error here: "The bucket you tried to delete is not empty. You must delete all versions in the bucket." But I get that error too with an unmodified version.

Would be nice to have some automated testing. I was thinking about spinning up Ceph in a Docker container and running a few tests against it. I've that for one of [my projects](https://travis-ci.org/pgerber/lo-migrate/builds/220872689) too and it appears to work pretty well. Also allows testing against [different versions of Ceph](https://travis-ci.org/pgerber/lo-migrate/builds/220872689). I might just write a .travis.yml assuming I find some time to do so and you'd be willing to enable Travis.
